src/lib389/lib389/ldapi.py

```diff
@@ -0,0 +1,50 @@
+# --- BEGIN COPYRIGHT BLOCK ---
+# Copyright (C) 2020 Red Hat, Inc.
+# All rights reserved.
+#
+# License: GPL (version 3 or any later version).
+# See LICENSE for details.
+# --- END COPYRIGHT BLOCK ---
+
+import ldap
+from lib389._mapped_object import DSLdapObject
+
+class LDAPIMapping(DSLdapObject):
+    """
+    Manage LDAPI mapping entries
+    """
+    def __init__(self, instance, container_dn, dn=None):
+        """
+        @param instance - a DirSrv instance
+        @param container_dn - the subtree where the mapping entries are located
+        """
+        super(LDAPIMapping, self).__init__(instance, dn)
+        self._dn = dn
+        self._rdn_attribute = 'cn'
+        self._filterattrs = ['cn']
+        self._container_dn = container_dn
+
+    def createMappingContainer(self):
+        """
+        Create the container entry
+        """
+        self._create_objectclasses = ['top', 'nsContainer']
+        parts = ldap.explode_dn(self._container_dn)
+        rdn, rdn_val = parts[0].split('=', 1)
+        basedn = ','.join(parts[1:])
+        self._dn = None
+        self.create(basedn=basedn, properties={'cn': rdn_val})
+
+    def createMappingEntry(self, name, username, ldap_dn):
+        """
+        @param name - name for the container entry.  This will be the RDN value.
+        @param username - the system username
+        @param ldap_dn - the LDAP DN that the system user should be mapped to
+        """
+        self._must_attributes = ['nsslapd-ldapiusername', 'nsslapd-authenticateasdn']
```

Comment:
Shouldn't these must attributes be defined in __init__ so that .create works without this helper call?

Reply:
Same comment :)

Reply:
No, because they are reset in the fixed mapping function.  I guess that means it should really be two classes I suppose :-/

Reply:
Yeah i think it's two classes.